### PR TITLE
feat(ModuleFinder): add filter for courses with exam clashes

### PR DESCRIPTION
**Pull request by TanLeYang**
## Context
#1184 
Additional filter option in module finder to hide modules with exams that clash with those of already selected modules

## Implementation
Added an additional filter checkbox in the module finder sidebar under the "Exams" section to allow users to filter out 
modules whose exams clash with those they have already added to their timetable.
Screenshot (20)

Implemented by connecting `ModuleFinderSidebar` to redux store using the `useSelector` hook and getting the active semester, modules already added to timetable and their semester data. From there, use the exam timings of the currently selected modules to generate `FilterItem` to be used by the ChecklistFilter.

## Other Information
This is my first time opening a PR to an open source project, please let me know if there's anything I'm doing wrongly!! 🙏

### CODE REVIEW

**Comment by TanLeYang** [comment]
I've fixed the lint issues!

File: website/src/views/modules/ModuleFinderSidebar.tsx [hunk 1]
```diff
@@ -40,6 +48,37 @@ const EXAM_FILTER_ITEMS: FilterItem[] = [
...
+    } = state;
+    const { timetable } = getSemesterTimetable(activeSemester, timetables);
+    const allSemesterModules = getSemesterModules(timetable, modules);
+    return allSemesterModules.map((module) => getModuleSemesterData(module, activeSemester));
+  });
+
+  const generateExamDateClashFilter: () => FilterItem[] = () => {
+    const examDates = selectedModules.map((module) => {
+      return module?.examDate;
+    });
```
Comment (ZhangYiJiang) [inline website/src/views/modules/ModuleFinderSidebar.tsx:66]:
```suggestion
    const examDates = selectedModules.map((module) => module?.examDate);
```

File: website/src/views/modules/ModuleFinderSidebar.tsx [hunk 2]
```diff
@@ -40,6 +48,37 @@ const EXAM_FILTER_ITEMS: FilterItem[] = [
...
+
+  const selectedModules = useSelector((state: StoreState) => {
+    const {
+      app: { activeSemester },
+      moduleBank: { modules },
+      timetables,
+    } = state;
+    const { timetable } = getSemesterTimetable(activeSemester, timetables);
+    const allSemesterModules = getSemesterModules(timetable, modules);
+    return allSemesterModules.map((module) => getModuleSemesterData(module, activeSemester));
```
Comment (ZhangYiJiang) [inline website/src/views/modules/ModuleFinderSidebar.tsx:60]: If you can, create `utils/null.ts` and add this util function, 

``` ts
export function isNotNull<T>(t: T | undefined | null): t is T {
  return t != null;
}
``` 

then you can filter out nullish values 


```suggestion
    return allSemesterModules.map((module) => getModuleSemesterData(module, activeSemester)).filter(isNotNull);
```

File: website/src/views/modules/ModuleFinderSidebar.tsx [hunk 3]
```diff
@@ -40,6 +48,37 @@ const EXAM_FILTER_ITEMS: FilterItem[] = [
...
+        label: 'No Exam Clash with Currently Selected Modules',
+        filter: BoolMustNot(TermsQuery('semesterData.examDate', examDates)),
+      },
+    ];
+  };
+
+  const getExamFilters: () => FilterItem[] = () => {
+    const examDateClashFilter = generateExamDateClashFilter();
+    return EXAM_FILTER_ITEMS.concat(examDateClashFilter);
+  };
```
Comment (ZhangYiJiang) [inline website/src/views/modules/ModuleFinderSidebar.tsx:80]: This and `generateExamDateClashFilter` don't seem to need to be functions. You should be able to inline `generateExamDateClashFilter` here and use `useMemo` 

```suggestion
  const examFilters = useMemo(() => {
    const examDates = selectedModules.map((module) => module?.examDate);
    const examDateClashFilter = {
      key: 'no-exam-clash',
      label: 'No Exam Clash with Currently Selected Modules',
      filter: BoolMustNot(TermsQuery('semesterData.examDate', examDates)),
    };
    
    return [...EXAM_FILTER_ITEMS, examDateClashFilter];
  }, [selectedModules]);
```

**Review by ZhangYiJiang — COMMENTED** [review]
This looks good! Thanks for doing this. Just a few simple suggestions and we should be good to merge

**Review by ZhangYiJiang — APPROVED** [review]

**Review by zwliew — APPROVED** [review]
Looks great. Thanks a lot guys :)

**Pull request merged**